Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp

```diff
@@ -37,32 +38,231 @@ FunctionConstructor::~FunctionConstructor()
+
+            // ii. Let nextArgString be ? ToString(nextArg).
+            // iii. Set P to the string-concatenation of P, "," (a comma), and nextArgString.
+            parameters.append(TRY(next_arg.to_string(global_object)));
+
+            // iv. Set k to k + 1.
+        }
+        parameters_string = String::join(',', parameters);
+
+        // f. Let bodyArg be args[k].
+        body_arg = args[k];
     }
-    auto is_generator = kind == FunctionKind::Generator || kind == FunctionKind::AsyncGenerator;
-    auto is_async = kind == FunctionKind::Async || kind == FunctionKind::AsyncGenerator;
-    auto source = String::formatted("{}function{} anonymous({}\n) {{\n{}\n}}", is_async ? "async " : "", is_generator ? "*" : "", parameters_source, body_source);
-    auto parser = Parser(Lexer(source));
-    auto function = parser.parse_function_node<FunctionExpression>();
-    if (parser.has_errors()) {
-        auto error = parser.errors()[0];
+
+    // 16. Let bodyString be the string-concatenation of 0x000A (LINE FEED), ? ToString(bodyArg), and 0x000A (LINE FEED).
+    auto body_string = String::formatted("\n{}\n", body_arg.has_value() ? TRY(body_arg->to_string(global_object)) : "");
+
+    // 17. Let sourceString be the string-concatenation of prefix, " anonymous(", P, 0x000A (LINE FEED), ") {", bodyString, and "}".
+    // 18. Let sourceText be ! StringToCodePoints(sourceString).
+    auto source_text = String::formatted("{} anonymous({}\n) {{{}}}", prefix, parameters_string, body_string);
+
+    u8 parse_options = FunctionNodeParseOptions::CheckForFunctionAndName;
+    if (kind == FunctionKind::Generator || kind == FunctionKind::AsyncGenerator)
+        parse_options |= FunctionNodeParseOptions::IsGeneratorFunction;
+    if (kind == FunctionKind::Async || kind == FunctionKind::AsyncGenerator)
+        parse_options |= FunctionNodeParseOptions::IsAsyncFunction;
+
+    // 19. Let parameters be ParseText(! StringToCodePoints(P), parameterSym).
+    i32 function_length = 0;
+    auto parameters_parser = Parser { Lexer { parameters_string } };
+    auto parameters = parameters_parser.parse_formal_parameters(function_length, parse_options);
+
+    // 20. If parameters is a List of errors, throw a SyntaxError exception.
+    if (parameters_parser.has_errors()) {
+        auto error = parameters_parser.errors()[0];
+        return vm.throw_completion<SyntaxError>(global_object, error.to_string());
+    }
+
+    // 21. Let body be ParseText(! StringToCodePoints(bodyString), bodySym).
+    bool contains_direct_call_to_eval = false;
+    auto body_parser = Parser { Lexer { body_string } };
+    // Set up some parser state to accept things like return and yield.
+    body_parser.state().in_function_context = true;
+    if (parse_options & FunctionNodeParseOptions::IsGeneratorFunction)
+        body_parser.state().in_generator_function_context = true;
+    (void)body_parser.parse_function_body(parameters, kind, contains_direct_call_to_eval);
```

Comment:
Do we not need to set `await_expression_is_valid` for async functions here?

Reply:
Oh and yes that name is different because I eventually wanted to move toward mimicking the `[~Await] [+Yield]` like names instead of in_generator.

(Also since await can be valid outside async functions)

Reply:
Yep, missed that. Fixed.